src/tracing/on_disk.rs

```diff
@@ -0,0 +1,240 @@
+    Option<fmt::Layer<S, fmt::format::DefaultFields, fmt::format::Format, MakeWriter>>,
+    filter::EnvFilter,
+    S,
+>;
+
+pub fn new<S>(config: &mut Config) -> Logger<S>
+where
+    S: tracing::Subscriber + for<'span> registry::LookupSpan<'span>,
+{
+    let make_writer = if config.enabled {
+        match MakeWriter::file(&config.log_file) {
+            Ok(make_writer) => make_writer,
+            Err(err) => {
+                config.enabled = false;
+
+                eprintln!(
+                    "failed to enable loggin into {} log-file: {err}",
+                    config.log_file,
+                );
+
+                MakeWriter::Drain
+            }
+        }
+    } else {
+        MakeWriter::Drain
+    };
+
+    let filter = if config.enabled {
+        filter(config.log_level.as_deref().unwrap_or(""))
+    } else {
+        "off".into()
+    };
+
+    let layer = fmt::Layer::default()
+        .with_ansi(false)
+        .with_span_events(config.span_events.clone().into())
+        .with_writer(make_writer);
+
+    Some(layer).with_filter(filter)
+}
+
+#[derive(Debug, Default)]
+pub struct Update {
+    make_writer: Option<MakeWriter>,
+    filter: Option<filter::EnvFilter>,
+    span_events: Option<fmt::format::FmtSpan>,
+}
+
+impl Update {
+    pub fn from_diff(diff: &ConfigDiff) -> anyhow::Result<Self> {
+        let mut update = Self::default();
+        update.prepare(diff)?;
+        Ok(update)
+    }
+
+    fn prepare(&mut self, diff: &ConfigDiff) -> anyhow::Result<()> {
+        let enabled = diff.enabled.unwrap_or(false);
+
+        if !enabled {
+            self.make_writer = Some(MakeWriter::Drain);
+            self.filter = Some("off".into());
+
+            return Ok(());
+        }
+
+        if let Some(log_file) = &diff.log_file {
+            self.make_writer = Some(MakeWriter::file(log_file)?);
+        }
+
+        if let Some(log_level) = &diff.log_level {
+            self.filter = Some(filter(log_level.as_deref().unwrap_or("")));
+        }
+
+        if let Some(span_events) = diff.span_events.clone() {
+            self.span_events = Some(span_events.into())
+        }
+
+        Ok(())
+    }
+
+    // `apply` should *never* log anything or produce any `tracing` events!
+    pub fn apply<S>(self, logger: &mut Logger<S>) {
+        if let Some(make_writer) = self.make_writer {
+            *logger
+                .inner_mut()
+                .as_mut()
+                .expect("valid logger state")
+                .writer_mut() = make_writer;
+        }
+
+        if let Some(filter) = self.filter {
+            *logger.filter_mut() = filter;
+        }
+
+        if let Some(span_events) = self.span_events {
+            let mut layer = logger.inner_mut().take().expect("valid logger state");
+            layer = layer.with_span_events(span_events);
+            *logger.inner_mut() = Some(layer);
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum MakeWriter {
+    File(fs::File),
+    Drain,
+}
```

Comment:
Would it be possible to 'just not create' a writer instead of creating a drain? Or does that cause problems with typing?

Reply:
Unfortunately, not possible. We have to provide _some_ writer.